### DISCUSSION TITLE
How to filter by an array of specs

### DISCUSSION THREAD

**Discussion by mationai** [uwdata/arquero#244] [Q&A] [answered]
I want to provide an array of specs, like
```js
[
  { date: '2020-01-01', val: 'a' },
  { date: '2020-01-01', val: 'b' },
]
```

and map/loop filter aq data with these fields such that row's date is `<` spec.date and .val matches. How do I do this?

@jheer any hints?

thank you

**Comment by mationai** [comment]
Hi @jheer . Appreciate any help on this.

**Comment by jheer** [comment, marked as answer]
I think you want to perform a [semijoin](https://uwdata.github.io/arquero/api/verbs#semijoin) with a custom predicate. Here is one possibility:

```js
const specs = aq.from([
  { date: new Date('2020-01-01'), val: 'a' },
  { date: new Date('2020-01-01'), val: 'b' }
]);

table.semijoin(specs, (a, b) => a.date < b.date && op.equal(a.val, b.val))
```

Using a custom predicate will fallback to an O(n*m) nested loop join, so watch out if the product of the table sizes is huge...

**Reply by mationai** [reply]
Thank you @jheer. I went back to the task at hand and what I need is to perform an operation based on the spec matching. For eg.

```js
const specs = aq.from([
  { date: new Date('2020-01-01'), val: 'x', ratio: 2 },
  { date: new Date('2020-01-01'), val: 'y', ratio: 3 }
]);

// logic (b = spec):
if (a.date < b.date && op.equal(a.val, b.val))
  a.someCol = a.someCol * b.ratio
```

How does the code for this look like?
